src/relations/db.py

```diff
@@ -0,0 +1,440 @@
+            The described state of this unit. Can be 'standalone', 'master', or 'standby'.
+        """
+        if standbys == "":
+            return "standalone"
+        # TODO this doesn't ever return false. Revisit mastery once scaling is sorted, and check
+        # replicas return standby.
+        elif self.charm.unit.is_leader():
+            return "master"
+        else:
+            return "standby"
+
+    def _on_relation_departed(self, departed_event: RelationDepartedEvent):
+        """Handle db-relation-departed event.
+
+        Removes relevant information from pgbouncer config when db relation is removed. This
+        function assumes that relation databags are destroyed when the relation itself is removed.
+        """
+        logger.info("db relation removed - updating config")
+        logger.warning(
+            f"DEPRECATION WARNING - {self.relation_name} is a legacy relation, and will be deprecated in a future release. "
+        )
+
+        app_databag = departed_event.relation.data[self.charm.app]
+        unit_databag = departed_event.relation.data[self.charm.unit]
+
+        for databag in [app_databag, unit_databag]:
+            databag["allowed-units"] = self.get_allowed_units(departed_event.relation)
+
+    def _on_relation_broken(self, broken_event: RelationBrokenEvent):
+        """Handle db-relation-broken event.
+
+        Removes all traces of the given application from the pgbouncer config, and removes relation
+        user if its unused by any other relations.
+
+        This doesn't delete any tables so we aren't deleting a user's entire database with one
+        command.
+        """
+        app_databag = broken_event.relation.data[self.charm.app]
+        user = app_databag.get("user")
+        database = app_databag.get("database")
+
+        if not self.charm.backend_postgres and None in [user, database]:
+            # this relation was never created, so wait for it to be initialised before removing
+            # everything.
+            logger.warning(
+                f"backend relation not yet available - deferring {self.relation_name}-relation-broken event."
+            )
+            broken_event.defer()
+            return
+
+        cfg = self.charm.read_pgb_config()
+
+        # check database can be deleted from pgb config, and if so, delete it. Database is kept on
+        # postgres application because we don't want to delete all user data with one command.
+        delete_db = False
+        for relname in ["db", "db-admin"]:
+            for relation in self.charm.model.relations.get(relname, []):
+                if relation.id == broken_event.relation.id:
+                    continue
+                if relation.data.get(self.charm.app, {}).get("database") == database:
+                    # There's multiple applications using this database, so don't remove it until
+                    # we can guarantee this is the last one.
+                    delete_db = True
```

Comment:
Should it really be deleted if there are more than one relation using it? The above comment says that we shouldn't delete the database.

Reply:
Good catch, this has been updated